be/src/exec/tablet_sink.cpp

```diff
@@ -172,138 +191,276 @@ Status NodeChannel::open_wait() {
+        _add_batch_counter.add_batch_wait_lock_time_us += closure->result.wait_lock_time_us();
+        _add_batch_counter.add_batch_num++;
+    }
+
+    for (auto& tablet : closure->result.tablet_vec()) {
+        TTabletCommitInfo commit_info;
+        commit_info.tabletId = tablet.tablet_id();
+        commit_info.backendId = _node_id;
+        std::vector<std::string> invalid_dict_cache_columns;
+        for (auto& col_name : tablet.invalid_dict_cache_columns()) {
+            invalid_dict_cache_columns.emplace_back(col_name);
+        }
+        commit_info.__set_invalid_dict_cache_columns(invalid_dict_cache_columns);
+
+        std::vector<std::string> valid_dict_cache_columns;
+        for (auto& col_name : tablet.valid_dict_cache_columns()) {
+            valid_dict_cache_columns.emplace_back(col_name);
+        }
+        commit_info.__set_valid_dict_cache_columns(valid_dict_cache_columns);
+
+        _tablet_commit_infos.emplace_back(std::move(commit_info));
     }
 
-    _eos_is_produced = true;
     return Status::OK();
 }
 
-Status NodeChannel::close_wait(RuntimeState* state) {
-    auto st = none_of({_cancelled, !_eos_is_produced});
-    if (!st.ok()) {
-        return _err_st;
+Status NodeChannel::_wait_all_prev_request() {
+    SCOPED_TIMER(_parent->_wait_response_timer);
+    if (_next_packet_seq == 0) {
+        return Status::OK();
     }
+    for (auto closure : _add_batch_closures) {
+        RETURN_IF_ERROR(_wait_request(closure));
+    }
+
+    return Status::OK();
+}
 
-    // waiting for finished, it may take a long time, so we could't set a timeout
-    MonotonicStopWatch timer;
-    timer.start();
-    while (!_add_batches_finished && !_cancelled) {
-        SleepFor(MonoDelta::FromMilliseconds(1));
+bool NodeChannel::_check_prev_request_done() {
+    if (UNLIKELY(_next_packet_seq == 0)) {
+        return true;
     }
-    timer.stop();
-    VLOG(1) << name() << " close_wait cost: " << timer.elapsed_time() / 1000000 << " ms";
 
-    if (_add_batches_finished) {
-        {
-            std::lock_guard<std::mutex> lg(_pending_batches_lock);
-            CHECK(_pending_chunks.empty()) << name();
-            CHECK(_cur_chunk == nullptr) << name();
+    for (size_t i = 0; i < _max_parallel_request_size; i++) {
+        if (_add_batch_closures[i]->count() == 1) {
+            _current_request_index = i;
+            return true;
         }
-        state->tablet_commit_infos().insert(state->tablet_commit_infos().end(),
-                                            std::make_move_iterator(_tablet_commit_infos.begin()),
-                                            std::make_move_iterator(_tablet_commit_infos.end()));
+    }
+
+    return false;
+}
+
+Status NodeChannel::_wait_one_prev_request() {
+    SCOPED_TIMER(_parent->_wait_response_timer);
+    if (_next_packet_seq == 0) {
+        return Status::OK();
+    }
+
+    // 1. check last request unblocking
```

Comment:
1. check the last request using short-circuit
This may be a better description.